lib/dialyzer/src/erl_types.erl

```diff
@@ -4410,6 +4410,11 @@ from_form({type, _Anno, 'fun', [{type, _, product, Domain}, Range]},
   {Dom1, L1, C1} = list_from_form(Domain, S, D, L, C),
   {Ran1, L2, C2} = from_form(Range, S, D, L1, C1),
   {t_fun(Dom1, Ran1), L2, C2};
+from_form({type, _Anno, 'fun', [{user_type, _, product, Domain}, Range]},
```

Comment:
We could add on a guard clause to reduce the foot print and duplication of the bodies, but I went with rule of three.